crates/bin/pd/src/main.rs

```diff
@@ -478,39 +478,34 @@ async fn main() -> anyhow::Result<()> {
                 )));
             }
 
-            let grpc_server = if let Some(domain) = grpc_auto_https {
-                use pd::auto_https::Wrapper;
-                use rustls_acme::{caches::DirCache, AcmeConfig};
-                use tokio_stream::wrappers::TcpListenerStream;
-                use tokio_util::compat::{FuturesAsyncReadCompatExt, TokioAsyncReadCompatExt};
-
-                let mut acme_cache = pd_home.clone();
-                acme_cache.push("rustls_acme_cache");
-
-                let bound_listener = TcpListener::bind(grpc_bind)
-                    .await
-                    .context(format!("Failed to bind HTTPS listener on {}", grpc_bind))?;
-                let listener = TcpListenerStream::new(bound_listener);
-                // Configure HTTP2 support for the TLS negotiation; we also permit HTTP1.1
-                // for backwards-compatibility, specifically for grpc-web.
-                let alpn_config = vec!["h2".into(), "http/1.1".into()];
-                let tls_incoming = AcmeConfig::new([domain.as_str()])
-                    .cache(DirCache::new(acme_cache))
-                    .directory_lets_encrypt(true) // Use the production LE environment
-                    .incoming(listener.map_ok(|conn| conn.compat()), alpn_config)
-                    .map_ok(|incoming| Wrapper {
-                        inner: incoming.compat(),
-                    });
-
-                tokio::task::Builder::new()
-                    .name("grpc_server")
-                    .spawn(grpc_server.serve_with_incoming(tls_incoming))
-                    .expect("failed to spawn grpc server")
-            } else {
-                tokio::task::Builder::new()
-                    .name("grpc_server")
-                    .spawn(grpc_server.serve(grpc_bind))
-                    .expect("failed to spawn grpc server")
+            // Now we drop down a layer of abstraction, from tonic to axum.
+            //
+            // TODO(kate): this is where we may attach additional routes upon this router in the
+            // future. see #3646 for more information.
+            let router = grpc_server.into_router();
+            let make_svc = router.into_make_service();
+
+            // Now start the GRPC server, initializing an ACME client to use as a certificate
+            // resolver if auto-https has been enabled.
+            macro_rules! spawn_grpc_server {
+                ($server:expr) => {
+                    tokio::task::Builder::new()
+                        .name("grpc_server")
+                        .spawn($server.serve(make_svc))
+                        .expect("failed to spawn grpc server")
+                };
+            }
+            let grpc_server = axum_server::bind(grpc_bind);
```

Comment:
Weird that this `bind` call isn't fallible; binding to 443 will fail as a normal user. We're not careful to document things like the need to run `setcap CAP_NET_BIND_SERVICE,CAP_NET_ADMIN=+ep /path/to/pd`. That's a docs issue, but I'd prefer to trust the program to exit if it cannot possibly run as configured. We dug into this a bit in paired review, and discovered that "bind" is weirdly not a verb in this context. Chucking this on the pile of "let's revisit soon, in the context of pd error-handling overall."

Reply:
> discovered that "bind" is weirdly not a verb in this context. Chucking this on the pile of "let's revisit soon, in the context of pd error-handling overall."

definitely the kind of thing that i was thinking about when i suggested in a meeting this week that moving directly to hyper will likely be a worthwhile goal at some point.

alternately, newer versions of axum expose a `serve` function, but that (aiui) is blocked on the tonic work linked in the description above.

for now, i agree that we can carry onwards and handle this another day. thank you for digging into this with me!